optaweb-vehicle-routing-backend/src/main/java/org/optaweb/vehiclerouting/service/distance/DistanceMatrixImpl.java

```diff
@@ -20,33 +20,50 @@
 import java.util.Map;
 import java.util.concurrent.ConcurrentHashMap;
 
+import javax.enterprise.context.ApplicationScoped;
+import javax.inject.Inject;
+
 import org.optaweb.vehiclerouting.domain.Distance;
 import org.optaweb.vehiclerouting.domain.Location;
 import org.optaweb.vehiclerouting.service.location.DistanceMatrix;
 import org.optaweb.vehiclerouting.service.location.DistanceMatrixRow;
-import org.springframework.beans.factory.annotation.Autowired;
-import org.springframework.stereotype.Service;
 
-@Service
+@ApplicationScoped
 class DistanceMatrixImpl implements DistanceMatrix {
 
     private final DistanceCalculator distanceCalculator;
-    private final DistanceRepository distanceRepository;
     private final Map<Location, Map<Long, Distance>> matrix = new HashMap<>();
 
-    @Autowired
-    DistanceMatrixImpl(DistanceCalculator distanceCalculator, DistanceRepository distanceRepository) {
+    @Inject
+    DistanceMatrixImpl(DistanceCalculator distanceCalculator) {
         this.distanceCalculator = distanceCalculator;
-        this.distanceRepository = distanceRepository;
     }
 
     @Override
     public DistanceMatrixRow addLocation(Location newLocation) {
+        Map<Long, Distance> distancesToOthers = updateMatrixLazily(newLocation);
+        return locationId -> {
+            if (!distancesToOthers.containsKey(locationId)) {
+                throw new IllegalArgumentException(
+                        "Distance from " + newLocation
+                                + " to " + locationId
+                                + " hasn't been recorded.\n"
+                                + "We only know distances to " + distancesToOthers.keySet());
+            }
+            return distancesToOthers.get(locationId);
+        };
+    }
+
+    private Map<Long, Distance> updateMatrixLazily(Location newLocation) {
         // Matrix == distance rows.
         // We're adding a whole new row with distances from the new location to existing ones.
         // We're also creating a new column by "appending" a new cell to each existing row.
         // This new column contains distances from each existing location to the new one.
 
+        if (matrix.containsKey(newLocation)) {
```

Comment:
Can be changed to `computeIfAbsent`: https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#computeIfAbsent-K-java.util.function.Function-